### PR TITLE
Avoid unnecessary refresh in Entry widget.

### PR DESCRIPTION
**Pull request by Dorbmon**
### Description:

There are two ideas for reducing the unnecessary (Without doing deep modification to the internal packages). 

First is when Validate() is called, we check if the validate has refreshed the widget. If so, we won't refresh.

Second is that, we used to just call extra refresh to ensure the cursor to be at the correct position(#4396). We can avoid that extra refresh by replacing the "e.updateTextAndRefresh" with updateText and calling the callback function manually.

I still think introducing a flag to forbid the refresh is a better idea so that we can avoid refreshing without modifying too many codes and also, without writing too many no-refresh version helper functions. Although it is not that elegant :)
Fixes #(issue)
#4397 
### Checklist:

- [ ] Tests included.
- [X] Lint and formatter run with no errors.
- [X] Tests all pass.

### CODE REVIEW

File: widget/entry.go [hunk 1]
```diff
@@ -893,11 +894,13 @@ func (e *Entry) TypedRune(r rune) {
 	})
 	e.propertyLock.Unlock()
 
-	e.Validate()
+	if err := e.Validate(); err == nil || errors.Is(err, e.validationError) {
+		// Under this situation, the validate function won't call refresh, so do it here
```
Comment (Jacalz) [inline widget/entry.go:898]: You are hard-coding an assumption about the method you are calling. This is bound to be problematic when someone changes the method's behaviour and forgets to update all occurrences of this assumption. It will basically introduce code-rot in the long run.

**Review by Jacalz — COMMENTED** [review]
I won't request changes to block this just because I have a different opinion but I'll let someone else in the team decide. 

However, I am still under the impression that an unexported `e.validate()` that doesn't refresh would be a better option. The exported `Validate()` could then call the unexported version and then refresh.

**Comment by Jacalz** [comment]
Also, fyi, I don’t believe this solves the issue in question entirely. There are quite a few places where we are refreshing twice.

File: widget/entry.go (hunk 1 above)
Comment (Dorbmon) [reply widget/entry.go:898]: Yes. So introduce the validate() without refresh might be a better option.

**Comment by Dorbmon** [comment]
Now I added the validate() which won't refresh and changed Validate() into a wrapper of validate() with refreshing.

**Comment by Dorbmon** [comment]
> Also, fyi, I don’t believe this solves the issue in question entirely. There are quite a few places where we are refreshing twice+ outside of validation and pasting.

Actually I think the best way to do the most optimization is to add the refresh barrier flag(maybe not a flag, can also be a counter, when it is 0, means no barrier). For example. the user might do some change with refresh in the callback function, you won't know if they will do that. Only if we set a barrier flag stopping the refresh function from really doing refresh, can we avoid the unnecessary refresh. Also, it is convenient because we can just set the barrier flag at the top level of the code.

File: widget/entry_validation.go [hunk 2]
```diff
@@ -27,13 +36,13 @@ func (e *Entry) SetOnValidationChanged(callback func(error)) {
 	e.onValidationChanged = callback
 }
 
-// SetValidationError manually updates the validation status until the next input change
-func (e *Entry) SetValidationError(err error) {
+// SetValidationError manually updates the validation status until the next input change. Returns if the validation error is really set.
+func (e *Entry) SetValidationError(err error) bool {
```
Comment (Jacalz) [inline widget/entry_validation.go:40]: You are breaking the public facing API here. Please revert this change.

**Review by Jacalz — CHANGES_REQUESTED** [review]
Looking better in my opinion. Though please note that you are currently breaking the stable API guarantee.

**Comment by Jacalz** [comment]
You might want to have a look at all the uses of `setFieldsAndRefresh()`. Saw a few more cases where refresh happens twice.

**Comment by Dorbmon** [comment]
> You might want to have a look at all the uses of `setFieldsAndRefresh()`. Saw a few more cases where refresh happens twice (`Undo() for example).

I believe I fix all this kind of duplicated refresh now.

File: widget/entry_validation.go (hunk 2 above)
Comment (Dorbmon) [reply widget/entry_validation.go]: OK

**Comment by Jacalz** [comment]
You might also wish to have a look at `ensureValidationSetup()`. I think it is only called when we refresh later.

File: widget/entry.go [hunk 3]
```diff
@@ -1218,14 +1230,15 @@ func (e *Entry) selectingKeyHandler(key *fyne.KeyEvent) bool {
...
+		e.setFields(e.eraseSelection)
 		e.propertyLock.RLock()
 		cb := e.OnChanged
 		e.propertyLock.RUnlock()
 		if cb != nil {
 			cb(e.Text)
 		}
-		e.Validate()
+		e.validate()
+		e.Refresh()
```
Comment (Jacalz) [inline widget/entry.go:1241]: You have move the refresh to before the callback in many other places. I suppose we want to be as consistent as possible. Not sure what is best but refreshing before potentially calling a long-running callback seems sensible?

**Comment by Dorbmon** [comment]
> You might also wish to have a look at `ensureValidationSetup()`. I think it is only called when we refresh later.

How about directly remove the Refresh() from ensureValidationSetup() as ensureValidationSetup() is only used inside the package.

**Comment by Jacalz** [comment]
> > You might also wish to have a look at `ensureValidationSetup()`. I think it is only called when we refresh later.
> 
> 
> 
> How about directly remove the Refresh() from ensureValidationSetup() as ensureValidationSetup() is only used inside the package.

That was my idea as well. Had a look through all of the Entry code looking at which unexported methods called Refresh() and then looked at where they were called.

File: widget/entry.go (hunk 3 above)
Comment (Dorbmon) [reply widget/entry.go:1241]: Yes. You are right.

Comment (Dorbmon) [reply widget/entry.go:1241]: But the validate() won't call the refresh. If so, we need to also move the validate. Is this consistent with the previous behavior? I found in the original code, sometimes we call the Validate() before the callback function ,but sometimes after it.

Comment (Jacalz) [reply widget/entry.go:1241]: I think it makes more sensor to validate before? The user might want to check the validation status maybe?

Comment (Jacalz) [reply widget/entry.go:1241]: I honestly don't know. We should be consistent though

**Comment by Jacalz** [comment]
Seems like there are a few test failures

**Comment by Dorbmon** [comment]
I got some problems about the test. I ran the test "Entry_SetValidationError" manually on my local machine(macOS). And I got the error:
```
Test:           TestEntry_SetValidationError
                Messages:       Image did not match master. Actual image written to file:///xxxx/fyne/widget/testdata/failed/entry/validation_set_valid.png.
```
But I checked the acutal image, it looks same to the standard image(file:///xxxx/fyne/widget/testdata/entry/validation_set_valid.png.)

**Comment by Jacalz** [comment]
Try to update the files (drag the files from the failed/ folder). There might be just some pixels in difference I think.

**Comment by Dorbmon** [comment]
I don't understand. I can pass the "Entry_SetValidationError" test on my local machine...

**Comment by andydotxyz** [comment]
> I don't understand. I can pass the "Entry_SetValidationError" test on my local machine...

Are you running on a M2 Mac by chance? They have somehow not the same image output to all other operating system.
So revert the image changes and see if the CI passes on this PR.
I still have to get to the bottom of why arm64 image output is different to all other arch...

**Comment by Dorbmon** [comment]
> > I don't understand. I can pass the "Entry_SetValidationError" test on my local machine...
> 
> Are you running on a M2 Mac by chance? They have somehow not the same image output to all other operating system. So revert the image changes and see if the CI passes on this PR. I still have to get to the bottom of why arm64 image output is different to all other arch...

Yes. I run on M2 pro chip. I've reverted the change but it still has some errors....

**Comment by andydotxyz** [comment]
> > > I don't understand. I can pass the "Entry_SetValidationError" test on my local machine...
> > 
> > 
> > Are you running on a M2 Mac by chance? They have somehow not the same image output to all other operating system. So revert the image changes and see if the CI passes on this PR. I still have to get to the bottom of why arm64 image output is different to all other arch...
> 
> Yes. I run on M2 pro chip. I've reverted the change but it still has some errors....

There is a test failure - the Entry border is no longer refreshing. When the state is invalid the border should be red... Comparing `form/hint_initial.png`

With your PR:
hint_initial

Before:
hint_initial

**Comment by Dorbmon** [comment]
But on my local machine, it works seems well? 
image
But still has error on my local machine...

**Comment by andydotxyz** [comment]
That isn't the same code. Run the unit test and check the output - you'll see that broken image I shared

**Comment by Jacalz** [comment]
I get the same result as @andydotxyz when running your code locally. I think there is a bug somewhere resulting in the border around the Entry not updating.

**Comment by Jacalz** [comment]
It would be great to try and get this landed soon. What I found strange though is that it seems to work fine when testing locally but, like noted above, the images for the Form widget seem to show that the border isn't updated correctly. Is there perhaps some Refresh missing somewhere?

**Comment by Dorbmon** [comment]
I don't have any idea how to solve this... I don't why the behavior is different between running on the local machine and the GitHub Action.

**Comment by Jacalz** [comment]
Andy, do you have any idea for what could be wrong? I've looked at the code but can't figure out what it can be.

**Comment by andydotxyz** [comment]
Test differing from the widget in real life is normally because refresh happens too late or misses something. The main UI refreshes a lot more due to user actions etc.
Given that this PR was removing refreshes it seems likely that this is the issue.

**Comment by Jacalz** [comment]
@Dorbmon Are you sure that you cannot replicate the test failures locally? The updated test files will be in `widget/testdata/failed` and its subfolders.

**Comment by Dorbmon** [comment]
@Jacalz I ran again on my local machine. I got some error:
Test:           TestForm_Disable_Validation
                Messages:       Image did not match master. Actual image written to file:///xxx/fyne/widget/testdata/failed/form/disable_validation_disabled_valid.png.
The follow image is the actual image:
image


But it looks different from CI's error:
Test:       	TestForm_Disable_Validation
        	Messages:   	Image did not match master. Actual image written to file:///home/runner/work/fyne/fyne/widget/testdata/failed/form/disable_validation_enabled_invalid.png.

**Comment by andydotxyz** [comment]
> [better name](https://github.com/fyne-io/fyne/pull/4465/commits/a1a11fcbe866abb15e6a17df180cfa5eb39dab94)

It looks like you have pushed a load of test debug code and/or commented out test cases.
With some of them disabled I can see why you might find that different lines of the test are failing - as it will stop at the first image that does not match.

**Comment by Jacalz** [comment]
@Dorbmon Would you have anything against me picking this up and trying to get the double refreshes resolved? I think it would be nice to land this in v2.4.5 but I also think it might be easier to start fresh and constantly run the tests to avoid running into these issues again :)

**Comment by Jacalz** [comment]
I opened #4680 for now. I can close it if you are not happy with me taking on the work.

**Comment by Dorbmon** [comment]
> I opened #4680 for now. I can close it if you are not happy with me taking on the work.

Just feel free to take this job. I am so busy these days and have no idea how to solve this:)

**Comment by Jacalz** [comment]
Alright. Thanks and thanks for working on this. You are always welcome back to working on PRs once you have more time :)
